Ignore deleted DRPC during MM activation

ELENAGER opened the pull request:
An unexpected activation of maintenance mode occurs when workloads are deleted after a failover operation.
It happens when DRPC is marked for deletion, but still exists. From now on maintenance mode activation is skipped for these DRPCs.

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2251381

nirs commented on controllers/drcluster_mmode.go:
```diff
@@ -59,6 +61,10 @@ func (u *drclusterInstance) mModeActivationsRequired() (map[string]ramen.Storage
 	for _, drpcCollection := range drpcCollections {
+		if !drpcCollection.drpc.GetDeletionTimestamp().IsZero() {
+			continue
+		}
```
Looks good, but `!drpcCollection.drpc.GetDeletionTimestamp().IsZero()` is not clear. It will be much easier to understand this if we add a helper like drClsuterIsDeleted(). We can actually add generic helper since all resources support GetDeletionTimestamp().

ELENAGER replied:
@nirs I added an issue to create generic helper for all resources: https://github.com/RamenDR/ramen/issues/1176

nirs commented:
@ELENAGER what is the flow when the unwanted activation happens? What is the event triggering the activation?

ELENAGER commented:
@nirs I am not fully acquainted with maintenance mode activation flow, but [here](https://github.com/RamenDR/ramen/pull/819) you can find more information

ELENAGER commented:
@ShyamsundarR , my first solution was to fix inside the `DRPCsFailingOverToClusterForPolicy` function, which is also called from `FilterDRCluster`, that filters DRPC resources that should be reconciled due to a DRCluster watch event. But then Nir and I had a long discussion and he thinks that the fact that DRPC is deleted does not mean we don't need to reconcile it for other reasons - we need to wait until the stuff is ready to complete the DRPC deletion. 
So, deleted DRPCs must be eliminated only from activate MM flow - and that's what I implemented

nirs commented:
Filtering deleted drpc in DRPCsFailingOverToClusterForPolicy means all callers will get this filter even it they are not interested in this filtering. Filtering deleted drcps in functions using for watching will skip reconcile of the drpc which may work in some cases or break others, or work now or break later. Filtering delete drpcs in the relevant code that need this filtering is clear and easier to understand and maintain.

ShyamsundarR commented:
Here is my reasoning to change `DRPCsFailingOverToCluster` to ignore deleted DRPC:

Callers of DRPCsFailingOverToCluster and why:
   - mModeActivationsRequired
     - Returns which storageID activations are required for DRPCs failing over to that cluster
       - If a deleted DRPC is ignored here, then that cluster activation would be pruned (iff there are no other DRPCs requiring the MMode)
  - FilterDRCluster
    - Gets DRPCs failing over to a DRCluster that underwent an Update event and issues a reconcile for them
      - If a deleted DRPC is passed in here for reconciliation, a deleted DRPC would be reconciled

We should ignore deleted DRPCs in both cases and need not specialize it for the case in mModeActivationsRequired only.

There would potentially be a problem in addition to the above solution as follows:
- If a DRPC was initiated to Failover to, say, `east` cluster
- Before MMode was active or the volumes were force promoted to Primary on `east`, the DRPC was deleted
- If the MMode is not retained for this corner case, then the delete MAY stall as it would wait for VRG to report a Primary, which MAY not happen as the MMode is not active

Now to fix the corner case as above, we could:
- Detect a DRPC as deleted, and if so ensure that if we have a VRG ManagedClusterView, that reports required primary state, and if not consider the DRPC for MMode activation etc.

The above would have a problem in garbage collecting MModes that are no longer required as the DRCluster is not watching and triggering reconcile for deleted DRPC (which was the original problem). For a deleted resource would would not get any resource that can help deteministically ensure which DRCluster should be considered for MMode pruning.

So overall, I think we can for now ignore this case, add a note in the code change that we are doing AND make this check in the DRPCsFailingOverToCluster so that deleted DRPCs are ignored for both cases (creating/pruning an MMode, reconciling DRPC due to changes in DRCluster of interest)

From the Bug the issue is as follows:

- When a DRPC is deleted its generation changes (++)
- Assuming this was a DRPC whose last spec.action was Failover, the check at [1] would detect that this is not yet failed over and hence trigger a MMode operation for this DRPC
- As the failover cluster would be the one where the workload is currently placed, the VRG fetched from this cluster is ignored at [2] and instead fetched from the S3 store (this per-se is not an issue, as we should be ignoring this DRPC rather than processing it)
- The result hence is to create the MMode for the required DRCluster
- Further, as DRCluster watcher for DRPC ignores delete events, a reconcile for DRCluster is not triggered, as that would have pruned the MMode request as there are no more DPRC requests in the queue [3]
- This results in leaving the MMode behind till the DRCluster is reconciled, or the ramen hub operator is restarted

From a fix perspective at [1] we potentially need to ignore DRPCs that are deleted. Further the DRCluster DRPC watcher should also process DRPC delete events, such that any outstanding MMode due to a deleted DRPC is garbage collected immediately.

[1] https://github.com/RamenDR/ramen/blob/d35f8bd8d73fb8e6fb0a90eccce30cbdb2b9309f/controllers/drplacementcontrol_controller.go#L455-L463
[2] https://github.com/RamenDR/ramen/blob/d35f8bd8d73fb8e6fb0a90eccce30cbdb2b9309f/controllers/drplacementcontrol.go#L616-L618
[3] https://github.com/RamenDR/ramen/blob/d35f8bd8d73fb8e6fb0a90eccce30cbdb2b9309f/controllers/drcluster_mmode.go#L34-L39

ELENAGER commented:
@ShyamsundarR thank you for the explanation. I fixed both in filtering deleted DRPCs and processing events.

BenamarMk commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -532,11 +532,17 @@ func DRPCsFailingOverToClusterForPolicy(
 
-	for idx, drpc := range drpcs.Items {
+	for idx := range drpcs.Items {
+		drpc := drpcs.Items[idx]
```
(nit) DRPC is passed by address in this function. Why don't you just not create a copy. I mean change this line to:
```
drpc := &drpcs.Items[idx]
```

ELENAGER replied:
Done, thanks!

ShyamsundarR commented on controllers/drcluster_controller.go:
```diff
@@ -175,7 +177,10 @@ func drpcPred() predicate.Funcs {
+			o := e.Object
+			log.Info("Processing DRPC delete event", "name", o.GetName(), "namespace", o.GetNamespace())
+
+			return true
```
I'm not sure about this, does delete actually have the resource that was deleted? The further calls here would go to `filterDRPC` and that would return the failoverCluster for reconciliation, assuming we have the object which on reading does not necessarily imply so.

I would ignore the delete in the watcher, or unconditionally trigger a reconcile for all DRClusters on a DRPC deletion detection.

Currently a stale MMode would be left behind if a failover was initiated, that caused an MMode to be set, and subsequently the DRPC was deleted before the failover was complete. We can note this for now, and not fix it as such.

ELENAGER replied:
We are doing exactly the same thing [here](https://github.com/RamenDR/ramen/blob/cae74b9e3eb2c742143a828767df9d6c7406798e/controllers/volumereplicationgroup_controller.go#L196), and then we are also filtering [here](https://github.com/RamenDR/ramen/blob/cae74b9e3eb2c742143a828767df9d6c7406798e/controllers/volumereplicationgroup_controller.go#L117-L129), so I assumed that we are actually having the resource, that was deleted

ShyamsundarR replied:
I am still not sure this helps us trap delete events of interest, we have the object type latch in `drpcMapFun` so should not cause panics or errors.

ELENAGER commented on controllers/drcluster_controller.go:
```diff
@@ -175,7 +177,10 @@ func drpcPred() predicate.Funcs {
 		},
 		DeleteFunc: func(e event.DeleteEvent) bool {
-			return false
+			o := e.Object
```
Done
